storage: make ReplicaEvalContext the interface

tbg opened the pull request:
As suggested by @bdarnell, this makes ReplicaEvalContext an interface
(previously `ReplicaI`) and in turn has it implemented by `*Replica` and
`SpanSetReplicaEvalContext` (formerly `ReplicaEvalContext`).

As fallout, some methods that previously returned errors due to assertions will
now Fatal. This was deemed more ergonomic than returning `nil` errors from
various `*Replica` accessors.

Introduce a second implementation, `*SpanSetReplicaEvalContext`, which is the
one we use in race-enabled builds (where we want to assert span accesses).

As a side effect, this also removes an oversight: we were previously passing the
`SpanSet` to the `ReplicaEvalContext` even when that wasn't desired (i.e.
outside of race builds).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19844)
<!-- Reviewable:end -->

nvanbenschoten commented:
Reviewed 10 of 11 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_eval_context_span.go, line 123 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI2kVBMnGTG14See2A:-KyI2kVBMnGTG14See2B:b-lqmtr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_eval_context_span.go#L123)):*
> ```Go
> // on Replica.ContainsKey.
> func (rec SpanSetReplicaEvalContext) ContainsKey(key roachpb.Key) bool {
> 	desc := rec.i.Desc()
> ```

Can we just call `rec.Desc`?

---

*[pkg/storage/replica_eval_context_span.go, line 157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI2yHI4J5duiTnz0oP:-KyI2yHI4J5duiTnz0oQ:b-5xubak) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_eval_context_span.go#L157)):*
> ```Go
> // transaction span GC.
> func (rec SpanSetReplicaEvalContext) GetTxnSpanGCThreshold() hlc.Timestamp {
> 	if rec.ss != nil {
> ```

It would be nice to avoid worrying about `nil` `*SpanSets` here. Perhaps we should create an `allSpans` `*SpanSet` and use that instead of ever using `nil`. Or we could change `SpanSet` to treat a nil receiver in `assertAllowed` as a no-op.

---

*[pkg/storage/replica_state.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3WNQhBT4pFe_DEcZ:-KyI3WNQhBT4pFe_DEc_:b-fib7t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_state.go#L163)):*
> ```Go
> }
> 
> // GetPushTxnQueue returns the Replica's pushTxnQueue.
> ```

This doesn't look right.

---

*[pkg/storage/replica_test.go, line 778 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3irAnyFVGjA1KPH9:-KyI3irAnyFVGjA1KPHA:b-5mhgps) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_test.go#L778)):*
> ```Go
> 		if _, err := evalRequestLease(context.Background(), tc.store.Engine(),
> 			CommandArgs{
> 				EvalCtx: &SpanSetReplicaEvalContext{tc.repl, nil},
> ```

Why aren't we using the constructor here and below?

---

*[pkg/storage/span_set.go, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3sKUumK-sDFkUIEO:-KyI3sKUumK-sDFkUIEP:bu40eqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/span_set.go#L103)):*
> ```Go
> }
> 
> // TODO(tschottdorf): take a context, make sure it's well-annotated.
> ```

And add a comment on this function please.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Updated, PTAL!

---

Review status: 4 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/storage/replica_eval_context_span.go, line 123 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI2kVBMnGTG14See2A:-KyI5hapfrNhtOlFC9Qk:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_eval_context_span.go#L123)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we just call `rec.Desc`?
</blockquote></details>

Good call, done.

---

*[pkg/storage/replica_eval_context_span.go, line 157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI2yHI4J5duiTnz0oP:-KyI5izyO23iil07fCXl:bew5zf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_eval_context_span.go#L157)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It would be nice to avoid worrying about `nil` `*SpanSets` here. Perhaps we should create an `allSpans` `*SpanSet` and use that instead of ever using `nil`. Or we could change `SpanSet` to treat a nil receiver in `assertAllowed` as a no-op.
</blockquote></details>

Good idea, done. In doing so, I discovered some sloppiness (https://github.com/cockroachdb/cockroach/issues/19851) that I won't be able to fix in this PR.

---

*[pkg/storage/replica_state.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3WNQhBT4pFe_DEcZ:-KyI4wYw0IqLR3VKdQ6G:b-f1vfyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_state.go#L163)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This doesn't look right.
</blockquote></details>

Already fixed.

---

*[pkg/storage/replica_test.go, line 778 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3irAnyFVGjA1KPH9:-KyI55jRRL1S6lsKl0xZ:buwwhcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/replica_test.go#L778)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why aren't we using the constructor here and below?
</blockquote></details>

No good reason, done.

---

*[pkg/storage/span_set.go, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyI3sKUumK-sDFkUIEO:-KyI5_9IrJGqcDocZOZ7:b-amf69g) ([raw file](https://github.com/cockroachdb/cockroach/blob/42e71513f110e1fdc22b9a3c42f023093edda69e/pkg/storage/span_set.go#L103)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

And add a comment on this function please.
</blockquote></details>

Done. I also removed the TODO because the key is good enough for finding out what failed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-:-KyIO_lyrH6tPTCF3bZQ:b-rr168o)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 11 files at r1, 6 of 6 files at r3.
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica_eval_context.go, line 50 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyIZ5eLZcvVu0ykKY5p:-KyIZ5eLZcvVu0ykKY5q:b-8gmk4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/36893cbc6684652c5a351af89ad55fbe8b15a46c/pkg/storage/replica_eval_context.go#L50)):*
> ```Go
> func NewReplicaEvalContext(r *Replica, ss *SpanSet) ReplicaEvalContext {
> 	if util.RaceEnabled {
> 		if ss == nil {
> ```

We should probably move this out of the `util.RaceEnabled` block so that any regressions are caught immediately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-:-KyIZISTo_Cjkp2Fc1d4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: make ReplicaEvalContext the interface

As suggested by @bdarnell, this makes ReplicaEvalContext an interface
(previously `ReplicaI`) and in turn has it implemented by `*Replica` and
`SpanSetReplicaEvalContext` (formerly `ReplicaEvalContext`).

As fallout, some methods that previously returned errors due to assertions will
now Fatal. This was deemed more ergonomic than returning `nil` errors from
various `*Replica` accessors.

Introduce a second implementation, `*SpanSetReplicaEvalContext`, which is the
one we use in race-enabled builds (where we want to assert span accesses).

As a side effect, this also removes an oversight: we were previously passing the
`SpanSet` to the `ReplicaEvalContext` even when that wasn't desired (i.e.
outside of race builds).

tschottdorf commented:
TFTR!

---

Review status: 8 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica_eval_context.go, line 50 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-KyIZ5eLZcvVu0ykKY5p:-KyI_8RUOhomMvZZwlGE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36893cbc6684652c5a351af89ad55fbe8b15a46c/pkg/storage/replica_eval_context.go#L50)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We should probably move this out of the `util.RaceEnabled` block so that any regressions are caught immediately.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19844#-:-KyI_7Bxs2D7MTDtHMr0:b-814g8k)*
<!-- Sent from Reviewable.io -->
